connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+        self._logger.info("Fetching all users")
+        self.usernames = {}
+        async for user in self.slack_client.list_users():
+            self.usernames[user["id"]] = self.get_username(user)
+            if self.configuration["sync_users"]:
+                yield self.adapt_user(user), None
+        async for message in self.channels_and_messages(self.usernames):
+            yield message, None
+
+    async def channels_and_messages(self, usernames):
+        """
+        Get all the messages from all the slack channels
+        :param usernames: a dictionary of userID -> username mappings, to facilitate replacing message references to userids with usernames
+        :return: generator for messages
+        """
+        delta = timedelta(days=self.n_days_to_fetch)
+        past_unix_timestamp = time.mktime((datetime.utcnow() - delta).timetuple())
+        current_unix_timestamp = time.mktime(datetime.utcnow().timetuple())
+        async for channel in self.slack_client.list_channels(
+            not self.auto_join_channels
+        ):
+            self._logger.info(f"Listed channel: '{channel['name']}'")
+            yield self.adapt_channel(channel, usernames)
+            channel_id = channel["id"]
+            join_response = {}
+            if self.auto_join_channels:
+                join_response = await self.slack_client.join_channel(
+                    channel_id
+                )  # can't get channel content if the bot is not in the channel
+            if not self.auto_join_channels or join_response.get("ok"):
+                async for message in self.slack_client.list_messages(
+                    channel_id, past_unix_timestamp, current_unix_timestamp
+                ):
+                    yield self.adapt_message(message, channel, usernames)
+            else:
+                self._logger.warning(
+                    f"Not syncing channel: '{channel['name']}' because: {join_response.get('error')}"
+                )
+
+    def get_username(self, user):
+        user_profile = user.get("profile", {})
+        if user_profile.get("display_name_normalized"):
+            return user_profile["display_name_normalized"]
+        elif user_profile.get("real_name_normalized"):
+            return user_profile["real_name_normalized"]
+        elif user.get("real_name"):
+            return user["real_name"]
+        elif user.get("name"):
+            return user["name"]
+        else:
+            return user["id"]
+
+    def adapt_user(self, user):
+        user_profile = user.get("profile", {})
+        user["real_name_normalized"] = user_profile.get("real_name_normalized")
+        user["profile_real_name"] = user_profile.get("real_name")
+        user["display_name"] = user_profile.get("display_name")
+        user["display_name_normalized"] = user_profile.get("display_name_normalized")
+        user["first_name"] = user_profile.get("first_name")
+        user["last_name"] = user_profile.get("last_name")
```

Comment:
There's not a consistent subset that's non-null. So I wanted to grab everything, so that there's _something_ stored, but I don't have to make a judgement call on which is preferred for search.